models/wiki_card_db.py

```diff
@@ -322,85 +324,289 @@ def clean_fields(cards, locales):
+        import pdb; pdb.set_trace()
+
+    if len(card_datas) == 1:
+        return card_datas
+
+    (
+        is_giant, merged_halves
+    ) = bifurcate_giants(card_datas)
+    if is_giant:
+        return merged_halves
+
+    (
+        has_anomaly, anomalies, other
+    ) = bifurcate_anomalies(card_datas)
+    if has_anomaly:
+        return anomalies + bifurcate_data(other)
+
+    (
+        has_redemption, redemption, other
+    ) = bifurcate_redemption(card_datas)
+    if has_redemption:
+        return redemption + bifurcate_data(other)
+
+    # Do special multi-house cases before this, like Redemption.
+    (
+        multi_house, merged
+    ) = bifurcate_multi_house(card_datas)
+    if multi_house:
+        return merged
+
+    # We shouldn't fall to this case and still have
+    # multiple types.
+    types = set([card["card_type"] for card in card_datas])
+    if len(types) > 1:
+        card_title = card_datas[0]["card_title"]
+        print(card_datas)
+        raise Exception(f"Unknown type mismatch {card_title} {types}")
+
+    # Otherwise the batch doesn't need bifurcation.
+    # TODO: what does it make here? Same card name from
+    # different sets? We build set data differently elsewhere?
+    #print("########################################")
+    #print("no bifurcate but multiples, by set?:")
+    #pprint.pp(card_datas)
+    #print("########################################")
+
     return card_datas
-                
 
-def build_json(only=None, build_locales=False):
+
+def bifurcate_giants(card_datas):
+    """Returns (is_giant, merged_halves)"""
+
+    card_title = card_datas[0]["card_title"]
+
+    texts = set([card["card_text"] for card in card_datas])
+
+    is_giant = False
+    for t in texts:
+        if "Play only with the other half of" in t:
+            is_giant = True
+            break
+
+    if not is_giant:
+        return (is_giant, [])
+
+    # Okay, giants. Separate by expansion and we expect
+    # have exactly two cards each. Then different expansions
+    # have different schema for finding the merged AA data.
+
+    same_exp = defaultdict(lambda: [])
+    for data in card_datas:
+        same_exp[data["expansion"]].append(data)
+
+    merged_halves = []
+    for _, batch in same_exp.items():
+        if len(batch) != 2:
+            pprint.pp(card_datas)
+            raise Exception(
+                "Giants weren't two halves per expansion?"
+            )
+
+        if batch[0]["card_type"] in ["Creature1", "Creature2"]:
+            # MM style
+            cre1 = None
+            cre2 = None
+            if batch[0]["card_type"] == "Creature1":
+                cre1 = batch[0]
+                cre2 = batch[1]
+            else:
+                cre1 = batch[1]
+                cre2 = batch[0]
+
+            new_data = {}
+            new_data.update(cre1)
+            new_data["card_type"] = "Creature"
+            merged_halves.append(new_data)
+
+        else:
+            # MoMu style
```

Comment:
Hah, they love to make things super confusing for us

Reply:
Keep us on our toes!